File: backend.spring/src/test/scala/com/skewwhiffy/auraltester/internalnotation/KeyFactoryTest.scala
package com.skewwhiffy.auraltester.internalnotation

import com.skewwhiffy.auraltester.testutils.{MockInstantiation, TestData}
import org.mockito.{InjectMocks, Mock}
import org.scalatest.flatspec.AnyFlatSpec
import org.scalatest.matchers.should

class KeyFactoryTest extends AnyFlatSpec with MockInstantiation {
  @Mock
  private val noteFactory: NoteFactory = null
  @InjectMocks
  private val keyFactory: KeyFactory = null

  it should "instantiate major key" in {
    val rawKey = "a#"
    val expectedNote = TestData.random.note
    when(noteFactory.getNote("A#")).thenReturn(expectedNote)

    val actual = keyFactory.getKey(rawKey)

    assert(actual.note == expectedNote)
    assert(!actual.isMinor)
  }

  it should "recognise m prefix for minor" in {
    val rawKey = "abm"
    val expectedNote = TestData.random.note
    when(noteFactory.getNote("Ab")).thenReturn(expectedNote)

    val actual = keyFactory.getKey(rawKey)

    assert(actual.note == expectedNote)
    assert(actual.isMinor)
  }

  it should "recognise minor suffix for minor" in {
    val rawKey = "bx minor"
    val expectedNote = TestData.random.note
    when(noteFactory.getNote("Bx")).thenReturn(expectedNote)

    val actual = keyFactory.getKey(rawKey)

    assert(actual.note == expectedNote)
    assert(actual.isMinor)
  }

}

File: backend/src/test/kotlin/com/skewwhiffy/notation/factory/KeyFactoryTest.kt
package com.skewwhiffy.notation.factory

import com.skewwhiffy.test.util.TestData
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.`when`
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class KeyFactoryTest {
  @Mock
  private lateinit var noteFactory: NoteFactory
  @InjectMocks
  private lateinit var keyFactory: KeyFactory

  @Test
  fun `instantiates major key`() {
    val rawKey = "a#"
    val expectedNote = TestData.random.note
    `when`(noteFactory.getNote("A#")).thenReturn(expectedNote)

    val actual = keyFactory.getKey(rawKey)

    assertThat(actual.note).isEqualTo( expectedNote)
    assertThat(actual.isMinor).isFalse
  }

  @Test
  fun `recognises m prefix for minor`() {
    val rawKey = "abm"
    val expectedNote = TestData.random.note
    `when`(noteFactory.getNote("Ab")).thenReturn(expectedNote)

    val actual = keyFactory.getKey(rawKey)

    assertThat(actual.note).isEqualTo(expectedNote)
    assertThat(actual.isMinor).isTrue
  }

  @Test
  fun `recognise minor suffix for minor`() {
    val rawKey = "bx minor"
    val expectedNote = TestData.random.note
    `when`(noteFactory.getNote("Bx")).thenReturn(expectedNote)

    val actual = keyFactory.getKey(rawKey)

    assertThat(actual.note).isEqualTo(expectedNote)
    assertThat(actual.isMinor).isTrue
  }
}